caching-service/src/main/java/org/zowe/apiml/caching/service/vsam/VsamStorage.java

```diff
@@ -1,151 +1,164 @@
-        Map<String, KeyValue> result = new HashMap<>();
-        List<VsamRecord> returned;
-
-        try (VsamFile file = new VsamFile(vsamConfig, VsamConfig.VsamOptions.READ)) {
-            returned = file.readForService(serviceId);
-        }
-
-        returned.forEach(vsamRecord -> result.put(vsamRecord.getKeyValue().getKey(), vsamRecord.getKeyValue()));
-
-        return result;
-    }
-
-}
+/*
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ */
+package org.zowe.apiml.caching.service.vsam;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.retry.annotation.Retryable;
+import org.zowe.apiml.caching.model.KeyValue;
+import org.zowe.apiml.caching.service.*;
+import org.zowe.apiml.caching.service.vsam.config.VsamConfig;
+import org.zowe.apiml.util.ObjectUtil;
+
+import java.util.*;
+
+/**
+ * Class handles requests from controller and orchestrates operations on the low level VSAM File class
+ */
+
+@Slf4j
+public class VsamStorage implements Storage {
+
+    private final VsamConfig vsamConfig;
+
+    public VsamStorage(VsamConfig vsamConfig, VsamInitializer vsamInitializer) {
+
+        log.info("Using VSAM storage for the cached data");
+
+        ObjectUtil.requireNotNull(vsamConfig.getFileName(), "Vsam filename cannot be null"); //TODO bean validation
+        ObjectUtil.requireNotEmpty(vsamConfig.getFileName(), "Vsam filename cannot be empty");
+        this.vsamConfig = vsamConfig;
+        log.info("Using Vsam configuration: {}", vsamConfig);
+        vsamInitializer.storageWarmup(vsamConfig);
+    }
+
+
+    @Override
+    @Retryable(value = {IllegalStateException.class, UnsupportedOperationException.class})
+    public KeyValue create(String serviceId, KeyValue toCreate) {
+        log.info("Writing record: {}|{}|{}", serviceId, toCreate.getKey(), toCreate.getValue());
+        KeyValue result = null;
+
+        try (VsamFile file = new VsamFile(vsamConfig, VsamConfig.VsamOptions.WRITE)) {
+
+            VsamRecord record = new VsamRecord(vsamConfig, serviceId, toCreate);
+            List<String> storedData = file.readRecords();
+            log.error("Size storedata: " + storedData.size());
+            verifyTotalSize(toCreate, storedData);
+            Optional<VsamRecord> returned = file.create(record);
+            if (returned.isPresent()) {
+                result = returned.get().getKeyValue();
+            }
+        }
+
+        if (result == null) {
+            throw new StorageException(Messages.DUPLICATE_KEY.getKey(), Messages.DUPLICATE_KEY.getStatus(), toCreate.getKey(), serviceId);
+        }
+
+        return result;
+    }
+
+    private void verifyTotalSize(KeyValue toCreate, List<String> storedData) {
```

Comment:
It's fine for now but as a part of the remove oldest strategy it would be good to use the same pattern that is now in the InMemory with EvictionStrategy as an interface implemented by the Reject and the RemoveOldest strategy.

Reply:
I agree, I can do the changes